### PR TITLE
Issue #2675 [staging & production] Can't mass-update permissions on child nodes

### PR DESCRIPTION
**Pull request by Johnetordoff**
# Purpose
This allows admins to mass update permissions on child nodes for contributors they've readded to the parent node of.

Closes issue https://github.com/CenterForOpenScience/osf.io/issues/2675

# Changes
Now when someone is added to a node they're already on there permission updates.

# Side Effects
None that I know of.

### CODE REVIEW

**Comment by chennan47** [comment]
@Johnetordoff Could you update the comment part according to the format requirement? Also close the pull request you replaced https://github.com/CenterForOpenScience/osf.io/pull/2914.

**Comment by GaryKriebel** [comment]
@Johnetordoff, please put the issue number this closes in the description and not in the title.

**Comment by samanehsan** [comment]
@Johnetordoff, could you take a look into the failing tests? Thanks!

**Comment by samanehsan** [comment]
This looks great! Just one small addition-- the `add_contributor` function will call `self.save()` if `save=True` is passed into the function: https://github.com/CenterForOpenScience/osf.io/blob/develop/website/project/model.py#L2237-#L2238. Can you add this in for when the permissions are overridden?

**Comment by samanehsan** [comment]
First pass complete! :+1:

File: website/project/model.py [hunk 1]
```diff
@@ -2239,6 +2239,11 @@ def add_contributor(self, contributor, permissions=None, visible=True,
 
             contributor_added.send(self, contributor=contributor, auth=auth)
             return True
+
+        #Permissions must be overridden if changed when contributor is added to parent he/she is already on a child of.
+        elif contrib_to_add in self.contributors and permissions is not None:
+            self.set_permissions(contrib_to_add, permissions, True)
```
Comment (samanehsan) [inline website/project/model.py]: This function has an optional `save` parameter with `False` as the default option so we do not want the default here to be that it always saves. Instead, if save is true, then save the node object. For example: https://github.com/CenterForOpenScience/osf.io/blob/develop/website/project/model.py#L2237-#L2238

Sorry if I wasn't clear earlier!

**Comment by samanehsan** [comment]
Can you add a test case for this change to `tests.test_models.TestProject`? That's where the other tests are for adding a contributor. After that, I think this should be good to go! :whale:

**Comment by Johnetordoff** [comment]
@samanehsan ping!

File: tests/test_models.py [hunk 2]
```diff
@@ -2525,6 +2525,31 @@ def test_replace_contributor(self):
             contrib.unclaimed_records.keys()
         )
 
+    def test_permission_update_on_readded_contributor(self):
+        # A user is added as a contributor
```
Comment (samanehsan) [inline tests/test_models.py]: I think this test is testing slightly different behavior (if a user gets added to a project with the correct permissions). Can you modify it to check that a contributor's permissions for a node get overridden when they are added from the parent project? Something like...
1) Add user to node with certain permissions
2) Add user to parent project with different permissions 
3) Assert that user node permissions are now the same as on the parent project.

**Comment by samanehsan** [comment]
Pass complete! :sunglasses: Please ping me when you're done making changes.

**Comment by Johnetordoff** [comment]
To my understanding, Issue  #2675 actually doesn't deal with that. It isn't for making the parent and child permissions consistent, it's specifically deals with re-adding people overwriting old child perms. I brought up the same "All-the-way-down" behavior in the comments here: #2675.

So I don't know how to formulate a test case for this. It's just created by a quirk of the UI. because what the issue really is, double dipping, using add_contributor for a node that a contrib is already on. In a perfect world this would not be allowed, but the UI is designed in a way that basically necessitates it.

File: tests/test_models.py [hunk 3]
```diff
@@ -2525,6 +2525,23 @@ def test_replace_contributor(self):
...
 
+    def test_permission_override_on_readded_contributor(self):
+
+        # A child node created
+        self.child_node = NodeFactory(parent=self.project, creator=self.consolidate_auth)
+
+        # A user is added as with read permission
+        user = UserFactory()
+
+        self.project.add_contributor(user, permissions=['read'])
```
Comment (samanehsan) [inline tests/test_models.py]: It's not necessary to add the user to the parent project for the test.

File: tests/test_models.py [hunk 4]
```diff
@@ -2525,6 +2525,23 @@ def test_replace_contributor(self):
...
+        self.child_node.add_contributor(user, permissions=['read'])
+
+        # user is readded with permission admin
+        self.child_node.add_contributor(user, permissions=['read','write','admin'])
+        self.project.save()
```
Comment (samanehsan) [inline tests/test_models.py]: Should this be `self.child_node.save()`?

**Comment by samanehsan** [comment]
Just a few more changes! :octopus:

**Comment by samanehsan** [comment]
Thanks! I think this is good to go @sloria! :octocat:

**Pull request merged**